ssl/s3_clnt.c

```diff
@@ -3090,17 +3108,23 @@ int ssl3_send_client_verify(SSL *s)
             n = j + 2;
         } else
 #endif
-        if (pkey->type == NID_id_GostR3410_2001) {
-            unsigned char signbuf[64];
+        if (pkey->type == NID_id_GostR3410_2001
+                || pkey->type == NID_id_GostR3410_2012_256
+                || pkey->type == NID_id_GostR3410_2012_512) {
+            unsigned char signbuf[128];
             int i;
-            size_t sigsize = 64;
-            s->method->ssl3_enc->cert_verify_mac(s,
-                                                 NID_id_GostR3411_94, data);
-            if (EVP_PKEY_sign(pctx, signbuf, &sigsize, data, 32) <= 0) {
+            size_t sigsize =
+                (pkey->type == NID_id_GostR3410_2012_512) ? 128 : 64;
+            int dgst_nid = NID_undef;
+
+            EVP_PKEY_get_default_digest_nid(pkey, &dgst_nid);
+            s->method->ssl3_enc->cert_verify_mac(s, dgst_nid, data);
+            if (EVP_PKEY_sign(pctx, signbuf, &sigsize, data,
+                              EVP_MD_size(EVP_get_digestbynid(dgst_nid))) <= 0) {
```

Comment:
Well, it is done at server side. In the request it is in ssl/s3_srvr.c, the part between lines 3034-3049